### PR TITLE
Feature/search warninglists

### PR DESCRIPTION
**Pull request by nikstuckenbrock**
https://drive.google.com/file/d/1q1rFpvIqMwb4JFqj7EchTcpQ9UQX3kYU/view?usp=sharing

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
We use warninglists in the MISP to track simple false positive lists e.g.. The reputation lookup commands only allow to exclude data that is in those lists. But we want to get a reputation for them with GOOD/BENIGN. Therefor I've added a new parameter which allows to search warninglists if no event was found with this data. The change should be downwards compatible. You can find a demo [here](https://forms.gle/XDfxU4E61ZwEESSMA). 

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/MISP/Integrations/MISPV3/MISPV3.py [hunk 1]
```diff
@@ -1800,6 +1876,7 @@ def main():
     malicious_tag_ids = argToList(params.get('malicious_tag_ids'))
     suspicious_tag_ids = argToList(params.get('suspicious_tag_ids'))
     benign_tag_ids = argToList(params.get('benign_tag_ids'))
+    search_warninglists: bool = argToBoolean(params.get('search_warninglists'))
```
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/MISPV3.py:1879]: please add here argToBoolean(params.get('search_warninglists', False)) - to have default value

File: Packs/MISP/Integrations/MISPV3/MISPV3.py
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/MISPV3.py:1]: Hi @nikstuckenbrock,
Thank you for your contribution!
Great work :)
Please see my comments below.
Thank you again

Comment (nikstuckenbrock) [reply Packs/MISP/Integrations/MISPV3/MISPV3.py:1]: Hi @tcarmeli1 ,

thanks for your feedback. I've applied your suggestions. :)

File: Packs/MISP/Integrations/MISPV3/README.md [hunk 2]
```diff
@@ -22,6 +22,7 @@ If you are upgrading from a previous version of this integration, see [Breaking
     | Malicious tag IDs | Comma-separated list of event's or attribute's malicious tag IDs. Malicious tags are stronger than suspicious tags. | False |
     | Suspicious tag IDs | Comma-separated list of event's or attribute's suspicious tag IDs. Malicious tags are stronger than suspicious tags. | False |
     | Benign tag IDs | Comma-separated list of event's or attribute's benign tag IDs. Malicious and suspicious tags are stronger than benign tags. | False |
+    | Search warninglists | Should the warninglists be considered when searching for an attribute | False |
```
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/README.md:25]: attribute or attributea?

**Review by tcarmeli1 — CHANGES_REQUESTED** [review]
Hi @nikstuckenbrock ,
Thank you for your contribution and great Demo.
Please see my newest comments.
Please also send a demo for the functionality when "search warninglists" is disabled.
Thank you again, Amazing work.

File: Packs/MISP/Integrations/MISPV3/README.md (hunk 2 above)
Comment (nikstuckenbrock) [reply Packs/MISP/Integrations/MISPV3/README.md:25]: attribute

**Review by tcarmeli1 — CHANGES_REQUESTED** [review]
Hi @nikstuckenbrock ,
Thank you for your amazing work!
Looks great.
Please send a short demo for the functionality when "search warninglists" is disabled.
Thank you so much!

File: Packs/MISP/Integrations/MISPV3/README.md (hunk 2 above)
Comment (ShirleyDenkberg) [inline Packs/MISP/Integrations/MISPV3/README.md:25]:
```suggestion
    | Search warninglists | Should the warninglists be considered when searching for an attribute.` | False |
```

File: Packs/MISP/ReleaseNotes/2_1_47.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+
+#### Integrations
+
+##### MISP v3
+
+Added the Search warninglists parameter in order to search in the warninglists for an attribute.
```
Comment (ShirleyDenkberg) [inline Packs/MISP/ReleaseNotes/2_1_47.md:6]:
```suggestion
Added the *Search warninglists* parameter in order to search in the warninglists for an attribute.
```

**Comment by ShirleyDenkberg** [comment]
@tcarmeli1 Doc review completed.

File: Packs/MISP/Integrations/MISPV3/MISPV3_test.py [hunk 4]
```diff
@@ -344,6 +344,30 @@ def test_convert_arg_to_misp_args(mocker):
     assert convert_arg_to_misp_args(args, args_names) == [{'dst-port': 8001}, {'src-port': 8002}, {'name': 'test'}]
 
 
+@pytest.mark.parametrize('dbot_type, dbot_score_type, value, expected_type_object', [
```
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/MISPV3_test.py:347]: Please add a test that shows the senario search_warninglists is true

**Review by tcarmeli1 — CHANGES_REQUESTED** [review]

File: Packs/MISP/Integrations/MISPV3/MISPV3_test.py (hunk 4 above)
Comment (nikstuckenbrock) [reply Packs/MISP/Integrations/MISPV3/MISPV3_test.py:347]: Sorry for the delayed response. I was ill the last days. I've added a corresponding test case.

Comment (tcarmeli1) [reply Packs/MISP/Integrations/MISPV3/MISPV3_test.py:347]: Hope you feel better now :)
Great work @nikstuckenbrock !

**Review by tcarmeli1 — CHANGES_REQUESTED** [review]
Hi @nikstuckenbrock ,
Hope you feel better now.
Great work!
Since this is Xsoar support PR- i would like to have a short demo meeting.
Please let me know your availability this week.
Feel free to contact me here or email tcarmeli@paloaltonetworks.com.
Thank you so much!

**Comment by tcarmeli1** [comment]
Scheduled demo to Wednesday at 9:30 UTC.

File: Packs/MISP/Integrations/MISPV3/MISPV3.py [hunk 5]
```diff
@@ -807,11 +828,53 @@ def get_indicator_results(
...
     else:
+        if search_warninglists:
+            values = argToList(value)
+            res: list = []
+            misp_warninglists_response = PYMISP.values_in_warninglist(values)
+            if 'errors' in misp_warninglists_response:
+                raise DemistoException(
+                    f'Unable to validate against MISP warninglists!\nError message: {misp_warninglists_response}')
+            if misp_warninglists_response and len(misp_warninglists_response.items()) > 0:
+                for _, lists in misp_warninglists_response.items():
```
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/MISPV3.py:841]: why are we using a for loop here? the loop will only happen once when we reach the return statement

File: Packs/MISP/Integrations/MISPV3/MISPV3.py [hunk 6]
```diff
@@ -807,11 +828,53 @@ def get_indicator_results(
...
+                        indicator=value,
+                        indicator_type=indicator_type,
+                        score=Common.DBotScore.GOOD, reliability=reliability,
+                        malicious_description=f"Match found in MISP warninglist{list_names}"
+                    )
+                    res.append(
+                        {
+                            "Value": value,
+                            "Count": len(lists),
+                            "Lists": ",".join([x["name"] for x in lists]),
```
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/MISPV3.py:853]: please use here the list_names variable

File: Packs/MISP/Integrations/MISPV3/MISPV3.py [hunk 7]
```diff
@@ -807,11 +828,53 @@ def get_indicator_results(
...
+                        }
+                    )
+                    human_readable = tableToMarkdown(
+                        "MISP Warninglist matchings:",
+                        sorted(res, key=lambda x: x["Count"], reverse=True),
```
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/MISPV3.py:858]: why do we need sorted here? res is a list with one dictionary value

**Review by tcarmeli1 — CHANGES_REQUESTED** [review]
Hi @nikstuckenbrock ,
Amazing demo, thank you so much.
Just some final comments.
Thank again.

File: Packs/MISP/Integrations/MISPV3/MISPV3.py (hunk 7 above)
Comment (nikstuckenbrock) [reply Packs/MISP/Integrations/MISPV3/MISPV3.py:858]: Sorting is not necessary here but it helps to keep the result deterministic. I can remove it if you want to.

File: Packs/MISP/Integrations/MISPV3/MISPV3.py (hunk 5 above)
Comment (nikstuckenbrock) [reply Packs/MISP/Integrations/MISPV3/MISPV3.py:841]: It is possible to search for multiple values therefor i need to loop through the results. But you're right, the return statement is at the wrong place here. I've moved it.

Comment (nikstuckenbrock) [reply Packs/MISP/Integrations/MISPV3/MISPV3.py]: You're right, the loop is not needed here. I've removed it.

Comment (tcarmeli1) [reply Packs/MISP/Integrations/MISPV3/MISPV3.py]: i think the for is needed but the return statement should be outside of the for loop, isn't it the case?
as i looked in both https://pymisp.readthedocs.io/en/latest/_modules/pymisp/api.html#PyMISP.values_in_warninglist and https://www.misp-project.org/openapi/#tag/Warninglists/operation/getWarninglistById and saw values_in_warninglists can get list of values and return a dictionary as seen here:
image
what do you think?

File: Packs/MISP/Integrations/MISPV3/MISPV3.py [hunk 8]
```diff
@@ -807,11 +828,57 @@ def get_indicator_results(
                               outputs_prefix='MISP.Attribute',
                               outputs_key_field='ID',
                               readable_output=readable_output)
+
     else:
+        if search_warninglists:
+            values = argToList(value)
```
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/MISPV3.py:834]: lets remove this value is a string

File: Packs/MISP/Integrations/MISPV3/MISPV3.py [hunk 9]
```diff
@@ -807,11 +828,57 @@ def get_indicator_results(
...
+            res: list = []
+            human_readable: str = ""
+            misp_warninglists_response = PYMISP.values_in_warninglist(values)
+            if 'errors' in misp_warninglists_response:
+                raise DemistoException(
+                    f'Unable to validate against MISP warninglists!\nError message: {misp_warninglists_response}')
+            if (misp_warninglists_response and isinstance(misp_warninglists_response, dict)
+                    and len(misp_warninglists_response.items()) > 0):
+                list_key: str = list(misp_warninglists_response.keys())[0]
```
Comment (tcarmeli1) [inline Packs/MISP/Integrations/MISPV3/MISPV3.py:843]: instead of lines 843+844 we can do lists = list(misp_warninglists_response.values())[0]

**Review by tcarmeli1 — CHANGES_REQUESTED** [review]
Hi @nikstuckenbrock ,
Thank you for your quick response, so after this change i have some additional comments.
Thank you so much!

File: Packs/MISP/Integrations/MISPV3/MISPV3.py (hunk 7 above)
Comment (tcarmeli1) [reply Packs/MISP/Integrations/MISPV3/MISPV3.py]: res is a list of one dictionary so we can remove the sorting

**Review by tcarmeli1 — APPROVED** [review]
Great work!

**Pull request merged**